README.md
# phenomap

This takes in a series of multi-layer raster files and returns a phenology projection raster.  `phenomap` is capable of analyzing satellite-derived NDVI and snowmelt time series on a regional or global scale.  Currently 'phenomap' is optimized for data that derives from the MOD 13 and MOD 10 data products, but it is expected that this project will eventually expand to also support GIMMS AVHRR and perhaps Landsat datasets.

## Motivation

Though landscape phenology is increasingly becoming a focal point of investigations of migration timing, hitherto no R packages existed that were able to reconstruct satellite-derived phenological metrics in space.  Thus, it was my motivation to develop this package *en passant* with John 2016 (M.S. thesis in ecology) in order to enable a broader group of researchers to project landscape timing measures in space.  

## Prerequisistes

`phenomap` requires at least R version 4.1.0, and has been tested on Mac OSX Monterey v12.3.1 and Windows 7.

This package uses methods from the `plyr` and `stringr`, and most importantly `phenex` and `terra` packages.  I recommend loading the `terra` package so that the product of `mapPheno()` can then be inspected and further processed as needed.

`phenomap` does not support .hdf files, and therefore it is recommended that users download and mosaic data.  I find that [pyModis](https://github.com/lucadelu/pyModis) is a reliable tool for such pursuits.  

## Warning

`phenomap` requires a considerable amount of computing power and even for small datasets it may take a while to process the data.  I therefore recommend cropping datasets to the minimum possible extent and projecting to the coarsest possible resolution.  At this point `phenomap` does not support parallel processing, but it is expected to in future versions.

## Installation

Use devtools to install this package...
```
devtools::install_github("JepsonNomad/phenomap")
```
... until it is available on CRAN, at which point install `phenomap` using
```
install.packages(phenomap)
```

## Code example
```
library(phenomap)
File_List.VI <- list.files(pattern="TeenyCrop") # these files can be found in the /inst/extdata folder
PhenoFactor = "VI"
threshold = 0.5
year = 2016
NDVI = 1
VIQ = 3
DOY = 4
PR = 5
verbose = TRUE
```
To project the timing of 50% GWI, use mapPheno with `phase = "greenup"`, and `threshold = 0.5`.
```
Sample.Greenup <- mapPheno(File_List = File_List.VI, PhenoFactor = PhenoFactor,
                           phase = "greenup", threshold = threshold, year = year,
                           NDVI = NDVI, VIQ = VIQ, DOY = DOY, PR = PR,
                           verbose = verbose)
plot(Sample.Greenup)
```

To project the timing of senescence, use `phase = "senescence"`:
```
Sample.Senescence <- mapPheno(File_List = File_List.VI, PhenoFactor = PhenoFactor,
                           phase = "senescence", threshold = threshold, year = year,
                           NDVI = NDVI, VIQ = VIQ, DOY = DOY, PR = PR,
                           verbose = verbose)
plot(Sample.Senescence)
```

To inspect broad measures of timing, compare histograms:

```
breaks<-seq(from=0,to=365,by=5)
hist(as.matrix(Sample.Senescence), breaks=breaks, ylim=c(0,150),
     col="ORANGE", main = "Snowmelt, Greenup, and Senescence",
     xlab="Day of Year")
hist(as.matrix(Sample.Greenup), breaks=breaks,
     col="palegreen", add=T)
```

To project the duration of the growing season, compare greenup and senescence rasters:

```
Growing.Season <- as.matrix(Sample.Senescence) - as.matrix(Sample.Greenup)
GS.raster <- rast(Growing.Season)
plot(GS.raster)
```

## License 

This project is licensed under the GPL-3 license.


NEWS.md
# phenomap 2.0.1

## Major changes

* Updated to support shift to terra.

## Minor changes

* Removed import of unnecessary libraries.


R/mapTrend.R
#' Convert a series of phenology terra::raster files to a single long-term trend terra::raster.
#'
#' @param File_List List of phenology terra::raster files (i.e. those produced in `mapPheno`)
#' @param Year_List Vector of Integer Year (YYYY) with length > 5
#' @param parallel TRUE or FALSE (Default = FALSE) if TRUE, use parallel backend through plyr::aaply
#' @param n.cores Integer number of cores to be used for parallel processing (only use if parallel = TRUE)
#' @param verbose TRUE or FALSE (Default = FALSE)
#' @return terra::raster object with extent=ext(rast(File_List)[1]) and CRS = crs(rast(File_List)[1]).  Layer 1 is the slope estimate of the linear model relating green-up timing (Day of Year) to time (Year).  Layer 2 is the p-value of the slope estimate.  Layer 3 is the standard error of the slope estimate.  Layer 4 is the r-squared value for the linear model.
#' @examples
#' \dontshow{
#' fpath <- system.file("extdata", package="phenomap")
#' File_List.Trend <- paste(fpath, list.files(path = fpath, pattern=c("Sample_Greenup_")), sep="/")[5:6]
#'
#' Year_List <- 2015:2016 # Tell it what years you're using
#' n.cores <- NA
#'
#' phenotrend <- mapTrend(File_List = File_List.Trend,
#'                              Year_List = Year_List,
#'                              parallel = FALSE,
#'                              n.cores = n.cores,
#'                              verbose=FALSE)
#'
#' }
#' \dontrun{
#'
#' fpath <- system.file("extdata", package="phenomap")
#' File_List.Trend <- paste(fpath, list.files(path = fpath, pattern=c("Sample_Greenup_")), sep="/")
#'
#' Year_List <- 2011:2016 # Tell it what years you're using
#' n.cores <- 4 # Set up parallel computing
#'
#' phenotrend <- mapTrend(File_List = File_List.Trend,
#'                              Year_List = Year_List,
#'                              parallel = TRUE,
#'                              n.cores = n.cores,
#'                              verbose=TRUE)
#'
#' }
#' @import stringr
#' @import terra
#' @importFrom plyr aaply
#' @importFrom doParallel registerDoParallel
#' @importFrom phenex phenoPhase
#' @export
# mapTrend added to phenomap v1.1.6

mapTrend <- function(File_List,
                     Year_List,
                     parallel = FALSE,
                     n.cores = NULL,
                     verbose = FALSE){
  if(parallel == TRUE){
    if(verbose){print("parallel option selected...")}
    if(verbose){print(paste0(n.cores, " cores selected"))}
    registerDoParallel(cores=n.cores)
    if(verbose){print("Parallel backend registered")}
  }

  myfiles <- File_List

  if(verbose){print(myfiles)}
  if(verbose){print(Sys.time())}

  phenostack <- terra::rast(lapply(X = myfiles,
                                   FUN = terra::rast))
  if(verbose){
    print("Stack created")
  }

  phenostack.array <- terra::as.array(phenostack)
  if(verbose){
    print(utils::str(phenostack.array))
  }

  # create function to determine pixel-specific linear trend through time series
  # requires minimum 5 non-NA cases
  trend.extraktor <- function(x){
    pixel.timeseries <- data.frame(x,Year_List)

    if(sum(!is.na(pixel.timeseries$x)) > 5){
      pixel.lm <- stats::lm(pixel.timeseries$x ~ pixel.timeseries$Year_List)
      ts.coeff <- summary(pixel.lm)$coefficients[2,1]

      extraktmat <- ts.coeff

      return(extraktmat)
    }
    else{return(NA)}
  }
  sig.extraktor <- function(x){
    pixel.timeseries <- data.frame(x,Year_List)

    if(sum(!is.na(pixel.timeseries$x)) > 5){
      pixel.lm <- stats::lm(pixel.timeseries$x ~ pixel.timeseries$Year_List)
      ts.sig <- summary(pixel.lm)$coefficients[2,4]

      extraktmat <- ts.sig

      return(extraktmat)
    }
    else{return(NA)}
  }
  se.extraktor <- function(x){
    pixel.timeseries <- data.frame(x,Year_List)

    if(sum(!is.na(pixel.timeseries$x)) > 5){
      pixel.lm <- stats::lm(pixel.timeseries$x ~ pixel.timeseries$Year_List)
      ts.se <- summary(pixel.lm)$coefficients[2,2]

      extraktmat <- ts.se

      return(extraktmat)
    }
    else{return(NA)}
  }
  R2.extraktor <- function(x){
    pixel.timeseries <- data.frame(x,Year_List)

    if(sum(!is.na(pixel.timeseries$x)) > 5){
      pixel.lm <- stats::lm(pixel.timeseries$x ~ pixel.timeseries$Year_List)
      ts.R2 <- summary(pixel.lm)$r.squared

      extraktmat <- ts.R2

      return(extraktmat)
    }
    else{return(NA)}
  }

  if(verbose){
    print("Trend extraktor created")
    print("Dumping regression metadata")
    print(Sys.time())
  }

  # extract the linear trend; dump to a trend array
  trend.array <- plyr::aaply(.data = phenostack.array,
                             .margins = c(1,2),
                             .fun = trend.extraktor,
                             .parallel = parallel)
  if(verbose){
    print("Trend data dumped")
    print(Sys.time())
  }

  sig.array <- plyr::aaply(.data = phenostack.array,
                           .margins = c(1,2),
                           .fun = sig.extraktor,
                           .parallel = parallel)
  if(verbose){
    print("Significance data dumped")
    print(Sys.time())
  }

  se.array <- plyr::aaply(.data = phenostack.array,
                          .margins = c(1,2),
                          .fun = se.extraktor,
                          .parallel = parallel)
  if(verbose){
    print("Standard error data dumped")
    print(Sys.time())
  }

  R2.array <- plyr::aaply(.data = phenostack.array,
                          .margins = c(1,2),
                          .fun = R2.extraktor,
                          .parallel = parallel)
  if(verbose){
    print("R^2 data dumped")
    print(Sys.time())
  }


  if(verbose){
    print("Regression metadata dumped")
    print("Converting trend.array to trend.raster")
    print(Sys.time())
  }

  trend.list <- list(trend.array,sig.array,se.array,R2.array)
  raster.list <- lapply(trend.list, terra::rast)
  trend.raster <- terra::rast(raster.list)

  names(trend.raster) <- c("Coefficient","P","Std.Error","R2")

  if(verbose){
    print("trend.raster created; function completed")
    print(Sys.time())
  }

  return(trend.raster)
}



R/mapPheno_v11.R
#' Convert a series of raster files to a single phenology raster.
#'
#' @param File_List List of raster files
#' @param PhenoFactor Character string; type of dataset to analyze (e.g., "VI", "Snow")
#' @param phase Character string; name of phenophase to be measured
#' @param threshold Float threshold GWI value to be projected
#' @param year Integer Year (YYYY)
#' @param NDVI Integer Band number of NDVI band in raster files
#' @param VIQ Integer Band number of VI Quality layer in raster files
#' @param DOY Integer Band number of Composite Day of Year layer in raster files
#' @param PR Integer Band Number of PR layer in raster files
#' @param SnowExtent Integer Band number of Maximum_Snow_Extent in raster files
#' @param verbose T/F verbose
#' @return Raster object with extent=extent(raster(File_List)[1]) and CRS = crs(raster(File_List)[1]).
#' @examples
#' list.files()
#' File_List <- list.files(pattern=c("TinyCrop_"))
#'
#' PhenoFactor = "VI"
#' phase = "greenup"
#' threshold = 0.05
#' year = 2016
#' NDVI = 1
#' VIQ = 3
#' DOY = 4
#' PR = 5
#' verbose = TRUE
#'
#' Sample.Greenup <- mapPheno(File_List = File_List, PhenoFactor = PhenoFactor,
#'                            phase = phase, threshold = threshold, year = year,
#'                            NDVI = NDVI, VIQ = VIQ, DOY = DOY, PR = PR,
#'                            verbose = verbose)
#'
#' @import dplyr
#' @import rgdal
#' @import stringr
#' @importFrom raster raster
#' @importFrom raster stack
#' @importFrom plyr aaply
#' @importFrom phenex modelNDVI
#' @importFrom phenex phenoPhase
#' @export

mapPheno <- function(File_List = NA, PhenoFactor = NA,
                     phase = NA, threshold = NA, year = NA,
                     NDVI = NA, VIQ = NA, DOY = NA, PR = NA,
                     SnowExtent = NA,
                     verbose = F){

  annualcrops <- File_List
  pos <- new.env()

  ## This function transforms a list of strings returned by searching for objects
  ## into a list of objects (required for proper stacking)
  if(verbose){print(paste0("Sorting data into 3-dimensional arrays... ", Sys.time()))}

  if(PhenoFactor == "VI"){

    if(verbose){print(paste0("VI option selected... ", Sys.time()))}

    if(is.na(NDVI) || is.na(VIQ) || is.na(DOY) || is.na(PR)){
      stop("Crucial Dataset missing (have you included NDVI, VIQ, DOY, and PR layers?)")
    }
    if(!is.na(NDVI)){
      if(verbose){print(paste0("Creating NDVI array... ", Sys.time()))}
      NDVI.Stack <- raster::stack(annualcrops, bands = NDVI)

      if(verbose){
        print(class(NDVI.Stack))
        print("NDVI.Stack dimensions ")
        print(dim(NDVI.Stack))
        print("2")
        }
      NDVI.Array <- as.array(NDVI.Stack)
      if(verbose){
        print(dim(NDVI.Array))
        print("3")
        }


      ## If NDVI is less than 0, it needs to be brought up to 0
      NDVIcleaner <- function(x) {x[x<0] <- 0; x};if(verbose){print("cleaner created")}
      NDVIcleaned <- apply(X = NDVI.Array, MARGIN = c(1,2,3), FUN = NDVIcleaner)
    };if(verbose){print("NDVI cleaned")}
    if(!is.na(VIQ)){
      if(verbose){print(paste0("Creating VI Quality array... ", Sys.time()))}
      VIQ.Stack <- stack(annualcrops, bands = VIQ)
      VIQ.Array <- as.array(VIQ.Stack); if(verbose){print("3")}

      first_k_bits <- function(int, k=16) {
        ## https://lpdaac.usgs.gov/products/modis_products_table/mod13q1 TABLE 2:
        ## MOD13Q1 VI Quality: "Bit 0 is the least significant (read bit words right to left)"
        integer_vector <- as.integer(intToBits(int))[1:k]
        return(paste(as.character(integer_vector), collapse=""))
      }

      VIQbinary <- apply(X = VIQ.Array, MARGIN = c(1,2,3), FUN = first_k_bits); if(verbose){print("4")}
    }
    if(!is.na(DOY)){
      if(verbose){print(paste0("Creating Day of Year array... ", Sys.time()))}
      DOY.Stack <- stack(annualcrops, bands = DOY); if(verbose){print("2")}
      DOY.Array <- as.array(DOY.Stack); if(verbose){print("3")}
    }
    if(!is.na(PR)){
      if(verbose){print(paste0("Creating Pixel Reliability array... ", Sys.time()))}
      PR.Stack <- stack(annualcrops, bands = PR)
      PR.Array <- as.array(PR.Stack)
    }

    # Clean NDVI data based on VIQ and PR data
    PRcleaner <- function(NDVISet, PRSet){
      Data <- NDVISet
      Meta <- PRSet
      for (i in 1:length(Data)){
        if (!is.na(Meta[i])){
          if (Meta[i] > 1){
            Data[i] <- NA
          }
          if(Meta[i] == -1){
            Data[i] <- NA
          }
        }
      }
      return(Data)
    }
    VIQcleaner <- function(NDVISet, VIQSet){
      Data <- NDVISet
      Meta <- VIQSet
      MODLAND.QA <- as.numeric(substring(Meta, 1, 2))
      VI.Usefulness <- as.numeric(substring(Meta, 3, 6))
      Aerosol.Quantity <- as.numeric(substring(Meta, 7, 8))
      Adjacent.Clouds <- as.numeric(substring(Meta, 9))
      BRDF.Correction <- as.numeric(substring(Meta, 10))
      Mixed.Clouds <- as.numeric(substring(Meta, 11))
      Land.Water.Flag <- as.numeric(substring(Meta, 12, 14))
      Possible.Snow.Ice <- as.numeric(substring(Meta, 15))
      Possible.Shadow <- as.numeric(substring(Meta, 16))
      for (i in 1:length(Data)){
        if (MODLAND.QA[i] > 5){
          Data[i] <- NA
        }
        if (VI.Usefulness[i] > 1100){
          Data[i] <- NA
        }
      }
      return(Data)
    }

    if(verbose){print(paste0("Cleaning NDVI data... ", Sys.time()))}

    FinalNDVI.PR <- PRcleaner(NDVISet = NDVIcleaned, PRSet = PR.Array)
    FinalNDVI.VIQ <- VIQcleaner(NDVISet = FinalNDVI.PR, VIQSet = VIQbinary)
    FinalNDVI <- FinalNDVI.VIQ

    ## Prep your NDVI values for Phenex processing
    ## Begin by creating a 3-D array in which
    ## NDVI values will are sorted by date of composite creation
    if(verbose){print(paste0("Sorting NDVI data into complete annual space-time cube... ", Sys.time()))}
    year_array <- array(data = NA, dim = c(nrow(FinalNDVI), ncol(FinalNDVI), 366))
    for (p in 1:dim(FinalNDVI)[3]){
      zindex <- p
      for (i in 1:nrow(FinalNDVI)){
        rowindex <- i
        for (j in 1:ncol(FinalNDVI)){
          colindex <- j
          if (!is.na(DOY.Array[rowindex, colindex, zindex])){
            year_array[rowindex, colindex, DOY.Array[rowindex, colindex, zindex]] <- FinalNDVI[rowindex,colindex, zindex]
          }
        }
      }
    }

    if(verbose){
      print(paste0("Converting annual space-time cube into phenoscape... ", Sys.time()))
      #print(year_array[120:130,230:240,42])
    }

    ## Create a matrix to be filled with phenology data
    phenodump <- array(data=NA, dim = c(nrow(FinalNDVI), ncol(FinalNDVI)))

    ## We want to apply the phenex functions through
    ## time in the FinalNDVI array;
    ## Meaning use aaply in dimensions c(1,2)
    NDVIFunction <- function(a){
      pixelID <- a
      ndvi <- modelNDVI((pixelID/10000), correction = "none",
                        method = "DLogistic", year=as.integer(year))[[1]]
      springtime <- phenoPhase(ndvi, phase = phase, method = "local", threshold = threshold)
      return(springtime)
    }


    if(verbose){print(paste0("Converting phenoscape to raster... ", Sys.time()))}
    phenodump <- aaply(.data = year_array,
                       .fun = NDVIFunction,
                       .margins = c(1,2),
                       .progress = "text")


    ## Turn final array back into a raster based on original raster properties
    phenoscape <- raster(x = phenodump,
                         template = NDVI.Objects.List[[1]])



  }

  if(PhenoFactor == "Snow"){
    if(is.null(SnowExtent)){
      stop("Crucial Dataset missing (have you included a SnowExtent layer?)")
    }
    if(!is.null(Extent.List)){
      Extent.Objects.List <- listobjecter(Extent.List)
      Extent.Stack <- stack(Extent.Objects.List)
      Extent.Array <- as.array(Extent.Stack)

      Complete.Extent.Array <- array(dim = c(dim(Extent.Array)[1], dim(Extent.Array)[2], 366))
      #Sort Extent.Array into daily slices (rather than 8-day slices)
      for(p in (1:dim(Extent.Array)[3])){
        zindex <- p
        zfinal <- p*8-7
        for(q in (1:nrow(Extent.Array))){
          rowindex <- q
          for(r in (1:ncol(Extent.Array))){
            colindex <- r
            Complete.Extent.Array[rowindex,colindex,zfinal] <- Extent.Array[rowindex,colindex,zindex]
          }
        }
      }

      # Create a phenodump array to deposit snowmelt phenology data
      phenodump.extent <- array(data=NA, dim = c(nrow(Complete.Extent.Array), ncol(Complete.Extent.Array)))

      # Fill in phenodump with the first date at which x == 25, that is,
      # the first time slice when pixels are detected as completely snow-free
      phenodump.extent <- apply(X = Complete.Extent.Array,
                                MARGIN = c(1,2),
                                FUN = function(x) min(which(x == 25)))

      phenoscape <- raster(x = phenodump.extent,
                           template = Extent.Objects.List[[1]])
    }

  }
  if(verbose) {print(Sys.time())}
  return(phenoscape)
}

#
# setwd("/Users/christianjohn/Documents/Post_Lab/Phenoscape/test")
# list.files()
# File_List <- list.files(pattern=c("TinyCrop_"))
# PhenoFactor = "VI"
# phase = "greenup"
# threshold = 0.5
# year = 2016
# NDVI = 1
# VIQ = 3
# DOY = 4
# PR = 5
# verbose = TRUE
#
# library(phenex)
# library(rgdal)
# library(raster)
# library(stringr)
# library(plyr)
# library(dplyr)
#
#
# Sample.Greenup <- mapPheno(File_List = File_List, PhenoFactor = PhenoFactor,
#                            phase = phase, threshold = threshold, year = year,
#                            NDVI = NDVI, VIQ = VIQ, DOY = DOY, PR = PR,
#                            verbose = verbose)
#


R/mapPheno.R
#' Convert a series of raster files to a single phenology raster.
#'
#' @param File_List List of raster files
#' @param PhenoFactor Character string; type of dataset to analyze (e.g., "VI", "Snow")
#' @param phase Character string; name of phenophase to be measured (e.g., "greenup", "snowmelt", "senescence" or other arguments passed to phenex::phenophase())
#' @param threshold Float threshold GWI value to be projected. Use only for VI option.
#' @param year Integer Year (YYYY)
#' @param NDVI Integer Band number of NDVI band in raster files
#' @param VIQ Integer Band number of VI Quality layer in raster files
#' @param DOY Integer Band number of Composite Day of Year layer in raster files
#' @param PR Integer Band Number of PR layer in raster files
#' @param SnowExtent Integer Band number of Maximum_Snow_Extent in raster files
#' @param verbose TRUE or FALSE (Default = FALSE)
#' @return Raster object with extent=extent(terra::rast(File_List)[1]) and CRS = crs(terra::rast(File_List)[1]).  Digital numbers are expressed as Day of Year.
#' @examples
#' \dontshow{
#' fpath <- system.file("extdata", package="phenomap")
#' File_List <- paste(fpath, list.files(path = fpath, pattern=c("ExCrop_")), sep="/")
#' File_List <- File_List[1:2]
#'
#' PhenoFactor = "VI"
#' phase = "greenup"
#' threshold = 0.5
#' year = 2016
#' NDVI = 1
#' VIQ = 3
#' DOY = 4
#' PR = 5
#' verbose = TRUE
#'
#' Sample.Greenup <- mapPheno(File_List = File_List, PhenoFactor = PhenoFactor,
#'                            phase = phase, threshold = threshold, year = year,
#'                            NDVI = NDVI, VIQ = VIQ, DOY = DOY, PR = PR,
#'                            SnowExtent=SnowExtent,
#'                            verbose = verbose)
#' }
#' \dontrun{
#' fpath <- system.file("extdata", package="phenomap")
#' File_List <- paste(fpath, list.files(path = fpath, pattern=c("TinyCrop_")), sep="/")
#' File_List
#'
#' PhenoFactor = "VI"
#' phase = "greenup"
#' threshold = 0.5
#' year = 2016
#' NDVI = 1
#' VIQ = 3
#' DOY = 4
#' PR = 5
#' verbose = TRUE
#'
#' Sample.Greenup <- mapPheno(File_List = File_List, PhenoFactor = PhenoFactor,
#'                            phase = phase, threshold = threshold, year = year,
#'                            NDVI = NDVI, VIQ = VIQ, DOY = DOY, PR = PR,
#'                            SnowExtent=SnowExtent,
#'                            verbose = verbose)
#' }
#' @import stringr
#' @import terra
#' @importFrom plyr aaply
#' @importFrom phenex modelNDVI
#' @importFrom phenex phenoPhase
#' @export

mapPheno<- function(File_List = NA, PhenoFactor = NA,
                    phase = NA, threshold = NA, year = NA,
                    NDVI = NA, VIQ = NA, DOY = NA, PR = NA,
                    SnowExtent = NA,
                    verbose = FALSE){

  annualcrops <- File_List
  pos <- new.env()

  ## This function transforms a list of strings returned by searching for objects
  ## into a list of objects (required for proper stacking)
  if(verbose){print(paste0("Sorting data into 3-dimensional arrays... ", Sys.time()))}

  if(PhenoFactor == "VI"){

    if(verbose){print(paste0("VI option selected... ", Sys.time()))}

    if(is.na(NDVI) || is.na(VIQ) || is.na(DOY) || is.na(PR)){
      stop("Crucial Dataset missing (have you included NDVI, VIQ, DOY, and PR layers?)")
    }
    if(!is.na(NDVI)){
      if(verbose){
        print(paste0("Creating NDVI array... ", Sys.time()))
        print("Array dimensions... ")
        print(c(dim(terra::as.array(terra::rast(annualcrops[1])))[1],
                dim(terra::as.array(terra::rast(annualcrops[1])))[2],
                length(annualcrops)))
      }
      NDVI.Array <- array(data=NA,
                          dim=c(dim(terra::as.array(terra::rast(annualcrops[1])))[1],
                                dim(terra::as.array(terra::rast(annualcrops[1])))[2],
                                length(annualcrops)),
                          dimnames=NULL)
      if(verbose){print("Created blank array")}
      for(i in 1:length(annualcrops)){
        NDVI.Array[,,i] <- terra::as.array(terra::rast(annualcrops[i], lyrs=NDVI))
      }; if(verbose){print("Filled array")}
      #NDVI.Stack <- stack(annualcrops, bands = NDVI)

      # if(verbose){
      #   print(class(NDVI.Stack))
      #   print(c("NDVI.Stack dimensions ", dim(NDVI.Stack)))
      # }
      #NDVI.Array <- as.array(NDVI.Stack)
      # if(verbose){
      #   print("Sample NDVI matrix dimensions")
      #   print(dim(terra::as.array(terra::rast(annualcrops[6], band=NDVI))))
      #   print("NDVI.Array dimensions")
      #   print(dim(NDVI.Array))
      #   print("3")
      # }


      ## If NDVI is less than 0, it needs to be brought up to 0
      NDVIcleaner <- function(x) {x[x<0] <- 0; x};if(verbose){print("cleaner created")}
      NDVIcleaned <- plyr::aaply(.data = NDVI.Array,
                                 .margins = c(1,2,3),
                                 .fun = NDVIcleaner)
    }
    if(verbose){print("NDVI cleaned")}
    if(!is.na(VIQ)){
      if(verbose){print(paste0("Creating VI Quality array... ", Sys.time()))}
      VIQ.Array <- array(dim=c(dim(terra::as.array(terra::rast(annualcrops[1])))[1],
                               dim(terra::as.array(terra::rast(annualcrops[1])))[2],
                               length(annualcrops)),
                         data=NA); if(verbose){print("Created blank array")}
      for(i in 1:length(annualcrops)){
        VIQ.Array[,,i] <- terra::as.array(terra::rast(annualcrops[i], lyrs=VIQ))
      }
      if(verbose){
        print("Filled VIQ Array")
        print("VIQ.Array dimensions")
        print(dim(VIQ.Array))
      }
      ## first_k_bits derived from:
      ## https://gis.stackexchange.com/questions/144441/how-can-i-parse-modis-mod13q1-quality-layers-in-r/144487
      first_k_bits <- function(int, k=16) {
        ## https://lpdaac.usgs.gov/products/modis_products_table/mod13q1 TABLE 2:
        ## MOD13Q1 VI Quality: "Bit 0 is the least significant (read bit words right to left)"
        integer_vector <- as.integer(intToBits(int))[1:k]
        return(paste(as.character(integer_vector),
                     collapse=""))
      }

      VIQbinary <- plyr::aaply(.data = VIQ.Array,
                               .margins = c(1,2,3),
                               .fun = first_k_bits); if(verbose){print("4")}
    }
    if(!is.na(DOY)){
      if(verbose){print(paste0("Creating Day of Year array... ", Sys.time()))}
      DOY.Array <- array(dim=c(dim(terra::as.array(terra::rast(annualcrops[1])))[1],
                               dim(terra::as.array(terra::rast(annualcrops[1])))[2],
                               length(annualcrops)),
                         data=NA)
      for(i in 1:length(annualcrops)){
        DOY.Array[,,i] <- terra::as.array(terra::rast(annualcrops[i], lyrs=DOY))
      }

    }
    if(!is.na(PR)){
      if(verbose){print(paste0("Creating Pixel Reliability array... ", Sys.time()))}
      PR.Array <- array(dim=c(dim(terra::as.array(terra::rast(annualcrops[1])))[1],
                              dim(terra::as.array(terra::rast(annualcrops[1])))[2],
                              length(annualcrops)),
                        data=NA)
      for(i in 1:length(annualcrops)){
        PR.Array[,,i] <- terra::as.array(terra::rast(annualcrops[i], lyrs=PR))
      }

    }

    # Clean NDVI data based on VIQ and PR data
    PRcleaner <- function(NDVISet, PRSet){
      Data <- NDVISet
      Meta <- PRSet
      for (i in 1:length(Data)){
        if (!is.na(Meta[i])){
          if (Meta[i] > 1){
            Data[i] <- NA
          }
          if(Meta[i] == -1){
            Data[i] <- NA
          }
        }
      }
      return(Data)
    }
    VIQcleaner <- function(NDVISet, VIQSet){
      Data <- NDVISet
      Meta <- VIQSet
      MODLAND.QA <- as.numeric(substring(Meta, 1, 2))
      VI.Usefulness <- as.numeric(substring(Meta, 3, 6))
      Aerosol.Quantity <- as.numeric(substring(Meta, 7, 8))
      Adjacent.Clouds <- as.numeric(substring(Meta, 9))
      BRDF.Correction <- as.numeric(substring(Meta, 10))
      Mixed.Clouds <- as.numeric(substring(Meta, 11))
      Land.Water.Flag <- as.numeric(substring(Meta, 12, 14))
      Possible.Snow.Ice <- as.numeric(substring(Meta, 15))
      Possible.Shadow <- as.numeric(substring(Meta, 16))
      for (i in 1:length(Data)){
        if (MODLAND.QA[i] > 10){
          Data[i] <- NA
        }
        if (VI.Usefulness[i] > 1100){
          Data[i] <- NA
        }
      }
      return(Data)
    }

    if(verbose){print(paste0("Cleaning NDVI data... ", Sys.time()))}

    #FinalNDVI.PR <- PRcleaner(NDVISet = NDVIcleaned, PRSet = PR.Array)
    FinalNDVI.VIQ <- VIQcleaner(NDVISet = NDVIcleaned, VIQSet = VIQbinary)
    FinalNDVI <- FinalNDVI.VIQ

    ## Prep your NDVI values for Phenex processing
    ## Begin by creating a 3-D array in which
    ## NDVI values will are sorted by date of composite creation
    if(verbose){
      print(paste0("Sorting NDVI data into complete annual space-time cube... ",
                   Sys.time()))
    }
    year_array <- array(data = NA, dim = c(nrow(FinalNDVI), ncol(FinalNDVI), 366))
    for (p in 1:dim(FinalNDVI)[3]){
      zindex <- p
      for (i in 1:nrow(FinalNDVI)){
        rowindex <- i
        for (j in 1:ncol(FinalNDVI)){
          colindex <- j
          if (!is.na(DOY.Array[rowindex, colindex, zindex])){
            if(DOY.Array[rowindex, colindex, zindex] > 0){
              year_array[
                rowindex, colindex, DOY.Array[
                  rowindex, colindex, zindex]] <-
                FinalNDVI[rowindex,colindex, zindex]
            }
          }
        }
      }
    }

    if(verbose){
      print(paste0("Converting annual space-time cube into phenoscape... ", Sys.time()))
      print("NDVI.Array pixel example")
      print(NDVI.Array[12,15,])
      print("VIQ Processed NDVI pixel example")
      print(FinalNDVI.VIQ[12,15,])
      print("Composite Day of Year pixel example")
      print(DOY.Array[12,15,])
      print("Annual space-time cube pixel example")
      print(year_array[12,15,])
    }

    ## Create a matrix to be filled with phenology data
    phenodump <- array(data=NA, dim = c(nrow(FinalNDVI), ncol(FinalNDVI)))

    ## We want to apply the phenex functions through
    ## time in the FinalNDVI array;
    ## Meaning use plyr::aaply in dimensions c(1,2)
    NDVIFunction <- function(a){
      pixelID <- a
      ndvi <- phenex::modelNDVI((pixelID/10000),
                                correction = "none",
                                method = "DLogistic",
                                year.int=as.integer(year))[[1]]
      springtime <- phenex::phenoPhase(ndvi,
                                       phase = phase,
                                       method = "local",
                                       threshold = threshold)[[1]]
      return(springtime)
    }

    phenodump <- plyr::aaply(.data = year_array,
                             .fun = NDVIFunction,
                             .margins = c(1,2),
                             .progress = "text")

    if(verbose){
      print("Phenoscape produced...")
      print(paste0("Converting phenoscape to raster... ", Sys.time()))
      print(class(phenodump))
      print(dim(phenodump))
    }

    ## Turn final array back into a raster based on original raster properties
    templateRast <- terra::rast(annualcrops[[1]],lyrs=1)
    phenoscape <- terra::rast(x = phenodump)
    terra::crs(phenoscape) <- terra::crs(templateRast)
    terra::ext(phenoscape) <- terra::ext(templateRast)
    # plot(phenoscape)
  }

  if(PhenoFactor == "Snow"){
    if(verbose){print("Snowmelt option selected...")}
    if(is.null(SnowExtent)){
      stop("Crucial Dataset missing (have you included a SnowExtent layer?)")
    }
    Extent.Stack <- terra::rast(annualcrops,
                                lyrs = SnowExtent)
    Extent.Array <- terra::as.array(Extent.Stack)

    Complete.Extent.Array <- array(dim = c(dim(Extent.Array)[1],
                                           dim(Extent.Array)[2],
                                           366))
    #Sort Extent.Array into daily slices (rather than 8-day slices)
    for(p in (1:dim(Extent.Array)[3])){
      zindex <- p
      zfinal <- p*8-7
      for(q in (1:nrow(Extent.Array))){
        rowindex <- q
        for(r in (1:ncol(Extent.Array))){
          colindex <- r
          Complete.Extent.Array[rowindex,colindex,zfinal] <-
            Extent.Array[rowindex,colindex,zindex]
        }
      }
    }

    # Create a phenodump array to deposit snowmelt phenology data
    phenodump.extent <- array(data=NA,
                              dim = c(nrow(Complete.Extent.Array),
                                      ncol(Complete.Extent.Array)))

    # Fill in phenodump with the first date at which x == 25, that is,
    # the first time slice when pixels are detected as completely snow-free
    phenodump.extent <- plyr::aaply(.data = Complete.Extent.Array,
                                    .margins = c(1,2),
                                    .fun = function(x) min(which(x == 25)))

    templateRast <- terra::rast(annualcrops[[1]],lyrs=1)
    phenoscape <- terra::rast(x = phenodump.extent)
    terra::crs(phenoscape) <- terra::crs(templateRast)
    terra::ext(phenoscape) <- terra::ext(templateRast)

  }
  if(verbose) {print(Sys.time())}
  return(phenoscape)
}

